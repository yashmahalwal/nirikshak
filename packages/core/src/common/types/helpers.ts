import { CustomFunction } from "./custom";
import { Literal, isLiteral } from "./literals";

export interface SchemaHelpers {
    [key: string]: CustomFunction;
}

// Ultimately, a resource fields ends in Primitive type
// Ex: {age: 23}
// Ex: {address: [{zipCode: "370445"}, {zipCode: "121342"}], classes: [1,2,3,4,5]}
export type Primitives = Literal | Array<Primitives>;
export function isPrimitives(input: any): input is Primitives {
    if (Array.isArray(input)) return input.every((i) => isPrimitives(i));
    return isLiteral(input);
}
