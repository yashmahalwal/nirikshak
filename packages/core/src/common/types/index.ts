export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
