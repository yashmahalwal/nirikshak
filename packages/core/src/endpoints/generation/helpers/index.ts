export { generateBodyType } from "./bodyTypeGen";
export { generateHeaderAndStatus } from "../output/headerAndStatus";
