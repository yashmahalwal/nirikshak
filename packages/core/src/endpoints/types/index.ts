export {
    MethodType,
    HeaderMap,
    Cases,
    BodyType,
    BodyInstance,
} from "./helpers";
export { isURLString } from "./urlString";
export { isResourceString } from "./resourceString";
export { isDescription } from "./description";
export { HeaderAndQuery, Inputs } from "./input";
export { HeaderAndStatus, Outputs } from "./output";
