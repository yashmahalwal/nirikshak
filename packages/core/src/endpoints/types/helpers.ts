import { ResourceString, isResourceString } from "./resourceString";
import { FakerType, isFakerType } from "../../common/types/fakerTypes";
import {
    CustomFunctionType,
    isCustomFunction,
} from "../../common/types/custom";
import { Literal, isLiteral } from "../../common/types/literals";
import { Primitives } from "../../common/types/helpers";

export type MethodType = "DELETE" | "GET" | "PUT" | "POST" | "PATCH";
export function isMethodType(key: any): key is MethodType {
    if (typeof key !== "string") return false;

    switch (key) {
        case "DELETE":
        case "GET":
        case "PUT":
        case "POST":
        case "PATCH":
            return true;
        default:
            return false;
    }
}

export type Cases = "NEGATIVE" | "POSITIVE" | "DESTRUCTIVE";

export function isCaseType(key: any): key is Cases {
    if (typeof key !== "string") return false;
    switch (key) {
        case "NEGATIVE":
        case "POSITIVE":
        case "DESTRUCTIVE":
            return true;
        default:
            return false;
    }
}

export type BaseType =
    | FakerType
    | CustomFunctionType
    | ResourceString
    | Literal;

export function isBaseType(input: any): input is BaseType {
    return (
        isFakerType(input) ||
        isCustomFunction(input) ||
        isResourceString(input) ||
        isLiteral(input)
    );
}

export type Base = BaseType | Base[];

export function isBase(input: any): input is Base {
    if (Array.isArray(input)) return input.every((value) => isBase(value));
    return isBaseType(input);
}

export type WithModifiers<T extends Base | BodyType> = (T extends Base
    ? { type: Base }
    : { field: BodyType }) & {
    plural?: boolean;
    nullable?: boolean;
    optional?: boolean;
};

// Checking the validity of the modifier fields of the HOT
function checkModifiers(input: WithModifiers<Base | BodyType>): boolean {
    let value = true;
    void (
        "nullable" in input &&
        (value = value && typeof input["nullable"] === "boolean")
    );
    void (
        "optional" in input &&
        (value = value && typeof input["optional"] === "boolean")
    );
    void (
        "plural" in input &&
        (value = value && typeof input["plural"] === "boolean")
    );

    return value;
}

// type guard: WithModifiers<ResourceBase>
export function isWithModifiersBodyType(
    input: any
): input is WithModifiers<BodyType> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("field" in input) {
        return isBodyType(input["field"]);
    }
    return false;
}

// type guard: WithModifies<BaseType>
export function isWithModifiersBase(input: any): input is WithModifiers<Base> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("type" in input) {
        return isBase(input["type"]);
    }
    return false;
}

// type guard: WithModifiers<BaseType> or WithModifiers<ResourceBase>
// Infers from input
export function isWithModifiers(
    input: any
): input is WithModifiers<Base | BodyType> {
    return isWithModifiersBase(input) || isWithModifiersBodyType(input);
}

export type OneOfEntries =
    | { fields: (BodyType | WithModifiers<BodyType>)[] }
    | { types: (Base | WithModifiers<Base>)[] }
    | {
          fields: (BodyType | WithModifiers<BodyType>)[];
          types: (Base | WithModifiers<Base>)[];
      };

// type guard: OneOfEntries
export function isOneOfEntries(input: any): input is OneOfEntries {
    if (!input || typeof input !== "object") return false;
    if (!Array.isArray(input["fields"]) && !Array.isArray(input["types"]))
        return false;

    if (Array.isArray(input["fields"]) && !input["fields"].length) return false;
    if (Array.isArray(input["types"]) && !input["types"].length) return false;

    let value = true;
    value =
        value &&
        (Array.isArray(input["types"])
            ? input["types"].every(
                  (entry) => isBase(entry) || isWithModifiersBase(entry)
              )
            : true);

    value =
        value &&
        (Array.isArray(input["fields"])
            ? input["fields"].every(
                  (entry) => isBodyType(entry) || isWithModifiersBodyType(entry)
              )
            : true);

    return value;
}

export interface BodyType {
    [key: string]:
        | ResourceString
        | Exclude<Base, ResourceString>
        | WithModifiers<Base>
        | WithModifiers<BodyType>
        | OneOfEntries
        | BodyType;
}

export interface BodyInstance {
    [key: string]: Primitives | BodyInstance | BodyInstance[];
}

export function isBodyType(input: any): input is BodyType {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (entry) =>
            isResourceString(entry) ||
            isBase(entry) ||
            isWithModifiers(entry) ||
            isOneOfEntries(entry) ||
            isBodyType(entry)
    );
}

export interface HeaderMap {
    [key: string]: Exclude<BaseType, null>;
}

export function isHeaderMap(input: any): input is HeaderMap {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (value) => value !== null && isBaseType(value)
    );
}
