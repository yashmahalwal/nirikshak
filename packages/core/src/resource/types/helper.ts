import {
    CustomFunctionType,
    isCustomFunction,
} from "../../common/types/custom";
import { Literal, isLiteral } from "../../common/types/literals";
import { FakerType, isFakerType } from "../../common/types/fakerTypes";
import { ResourceBase, isResourceBase } from "./resource";
import { Primitives } from "../../common/types/helpers";

// An instance is of the following type
export interface ResourceInstanceBase {
    [key: string]: Primitives | ResourceInstanceBase | ResourceInstanceBase[];
}

export interface ResourceInstance extends ResourceInstanceBase {
    id: string | number;
}

// Base type: Simple literals that a resource field can have
export type BaseType =
    | FakerType
    | CustomFunctionType
    | Literal
    | Array<BaseType>;

// Validation for the same
export function isBaseType(input: any): input is BaseType {
    if (Array.isArray(input)) return input.every((entry) => isBaseType(entry));

    // Ordering is important
    return isFakerType(input) || isCustomFunction(input) || isLiteral(input);
}

// Higher order type: Adding modifiers to a resource field entry
// A resource base field can be: Base Type or another Resource base
// You can add nullable, plural or optional here
// Ex: {type: "faker:random.words", nullable: true}
// Ex: {field: {name: "custom:myName"}, plural: true}
export type WithModifiers<
    T extends BaseType | ResourceBase
> = (T extends BaseType
    ? {
          type: BaseType;
      }
    : {
          field: ResourceBase;
      }) & { plural?: boolean; nullable?: boolean; optional?: boolean };

// Checking the validity of the modifier fields of the HOT
function checkModifiers(
    input: WithModifiers<ResourceBase | BaseType>
): boolean {
    let value = true;
    void (
        "nullable" in input &&
        (value = value && typeof input["nullable"] === "boolean")
    );
    void (
        "optional" in input &&
        (value = value && typeof input["optional"] === "boolean")
    );
    void (
        "plural" in input &&
        (value = value && typeof input["plural"] === "boolean")
    );

    return value;
}

// type guard: WithModifiers<ResourceBase>
export function isWithModifiersResource(
    input: any
): input is WithModifiers<ResourceBase> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("field" in input) {
        return isResourceBase(input["field"]);
    }
    return false;
}

// type guard: WithModifies<BaseType>
export function isWithModifiersBaseType(
    input: any
): input is WithModifiers<BaseType> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("type" in input) {
        return isBaseType(input["type"]);
    }
    return false;
}

// type guard: WithModifiers<BaseType> or WithModifiers<ResourceBase>
// Infers from input
export function isWithModifiers(
    input: any
): input is WithModifiers<BaseType> | WithModifiers<ResourceBase> {
    return isWithModifiersBaseType(input) || isWithModifiersResource(input);
}

// Selecting one from multiple choices
export type OneOfEntries =
    | {
          types: Array<BaseType | WithModifiers<BaseType>>;
      }
    | {
          fields: Array<ResourceBase | WithModifiers<ResourceBase>>;
      }
    | {
          fields: Array<ResourceBase | WithModifiers<ResourceBase>>;
          types: Array<BaseType | WithModifiers<BaseType>>;
      };

//   type guard: OneOfEntries
export function isOneOfEntries(input: any): input is OneOfEntries {
    if (!input || typeof input !== "object") return false;
    if (!Array.isArray(input["fields"]) && !Array.isArray(input["types"]))
        return false;

    if (Array.isArray(input["fields"]) && !input["fields"].length) return false;
    if (Array.isArray(input["types"]) && !input["types"].length) return false;

    let value = true;
    value =
        value &&
        (Array.isArray(input["types"])
            ? input["types"].every(
                  (entry) => isBaseType(entry) || isWithModifiersBaseType(entry)
              )
            : true);

    value =
        value &&
        (Array.isArray(input["fields"])
            ? input["fields"].every(
                  (entry) =>
                      isResourceBase(entry) || isWithModifiersResource(entry)
              )
            : true);

    return value;
}
