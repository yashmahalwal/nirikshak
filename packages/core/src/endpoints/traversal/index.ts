import Supertest from "supertest";
import { MethodType, Inputs, Outputs } from "../types";
import { URLString } from "../types/urlString";
import { ResourceInstance, Resource } from "../../resource";
import { SchemaHelpers } from "../../common";
import { Collection } from "./collection";
import { HeadersInstance } from "../generation/helpers/headerMapGen";
import { makeGetRequest } from "./get";

export { TraversalHelpers, TraversalHelperFunctions } from "./traversalHelpers";
export { Collection } from "./collection";

export async function makeRequest(
    url: URLString,
    method: MethodType,
    caseValue: keyof Outputs[typeof method],
    server: Supertest.SuperTest<Supertest.Test>,
    input: Inputs[typeof method],
    resource: ResourceInstance,
    helpers: SchemaHelpers,
    collection: Collection,
    resourceJSON: Resource
): Promise<{
    status: number;
    headers?: HeadersInstance;
    body?: any;
}> {

    switch(method){
        case "GET":
            return makeGetRequest(caseValue, server,url,input,resource,helpers,resourceJSON);
    }

    return { status: 200 };
}
