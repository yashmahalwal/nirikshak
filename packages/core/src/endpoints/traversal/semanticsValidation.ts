import { HeaderAndStatus } from "../types/output";
import { ResourceInstance } from "../../resource/types/helper";
import { SchemaHelpers } from "../../common/types/helpers";
import { matchBaseType } from "./bodyValidation";
import { HeaderMap } from "../types/helpers";

export async function headersValidation(
    input: any,
    headers: HeaderMap[],
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<boolean> {
    if (!input || typeof input !== "object") return false;

    return (
        !headers.length ||
        (
            await Promise.all(
                headers.map(async (header) =>
                    (
                        await Promise.all(
                            Object.keys(header).map((key) =>
                                matchBaseType(
                                    input[key],
                                    header[key],
                                    resource,
                                    helpers
                                )
                            )
                        )
                    ).every((val) => val)
                )
            )
        ).some((val) => val)
    );
}

export function statusValidation(input: any, expected: number[]): boolean {
    return !expected.length || expected.some((entry) => entry === input);
}

export function extractStatusFromSemantics(
    entry: { semantics: HeaderAndStatus }[]
): number[] {
    return entry.map((e) => e.semantics.status);
}

export function extractHeadersFromSemantics(
    entry: { semantics: HeaderAndStatus }[]
): HeaderMap[] {
    const o: HeaderMap[] = [];
    entry.forEach(
        (e) => "headers" in e.semantics && o.push(e.semantics.headers!)
    );
    return o;
}
