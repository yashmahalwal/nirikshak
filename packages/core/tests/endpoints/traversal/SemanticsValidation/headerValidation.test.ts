import { headersValidation } from "../../../../src/endpoints/traversal/semanticsValidation";
import { SchemaHelpers } from "../../../../src/common/types/helpers";
import { ResourceInstance } from "../../../../src/resource/types/helper";
import faker from "faker";
import { HeaderMap } from "../../../../src/endpoints/types/helpers";
const Helpers: SchemaHelpers = {
    word: (): Promise<string> =>
        new Promise((resolve) =>
            setTimeout(() => resolve(faker.random.word()), 10)
        ),
    async invalidHelper() {
        return { age: 10 } as any;
    },
};

const ValidResource: ResourceInstance = {
    created: true,
    id: faker.random.uuid(),
    args: [1, 2, 3, [false], true, "strstr", null],
    addresses: [{ zipCode: "370432" }, { zipCode: "232344" }],
    fallacies: faker.random.word(),
    places: {
        to: {
            be: {
                miles: {
                    to: {
                        go: {
                            before: 23,
                        },
                    },
                },
            },
        },
    },
};

const ValidInputs: {
    input: () => Promise<any>;
    headers: HeaderMap[];
    output: boolean;
}[] = [
    {
        input: async (): Promise<null> => null,
        headers: [],
        output: false,
    },
    {
        input: async (): Promise<string> => "TureHeader",
        headers: [],
        output: false,
    },
    {
        input: async (): Promise<Record<string, any>> => ({ h: true }),
        headers: [],
        output: true,
    },
    {
        input: async (): Promise<Record<string, any>> => ({}),
        headers: [],
        output: true,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            a: true,
            b: false,
        }),
        headers: [
            {
                a: true,
                b: false,
                c: "faker:random:number",
                d: "custom:word",
                e: "resource:id",
            },
        ],
        output: false,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            a: true,
            b: false,
            c: faker.random.number(),
            d: await Helpers.word(),
            e: ValidResource.id,
        }),
        headers: [
            {
                a: true,
                b: false,
                c: "faker:random:number",
                d: "custom:word",
                e: "resource:id",
            },
        ],
        output: false,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            a: true,
            b: false,
            c: faker.random.number(),
            d: await Helpers.word(),
            e: ValidResource.id,
        }),
        headers: [
            {
                a: true,
                b: false,
            },
        ],
        output: true,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            a: true,
            b: true,
            c: faker.random.number(),
            d: await Helpers.word(),
            e: ValidResource.id,
        }),
        headers: [
            {
                a: true,
                b: false,
            },
        ],
        output: false,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            a: true,
            b: true,
            c: faker.random.number(),
            d: await Helpers.word(),
            e: ValidResource.id,
        }),
        headers: [
            {
                a: true,
                b: false,
            },
            { a: true },
        ],
        output: true,
    },
    {
        input: async (): Promise<Record<string, any>> => ({}),
        headers: [
            { a: true },
            { b: "custom:word" },
            { c: "faker:random:word" },
            { d: "resource:id" },
        ],
        output: false,
    },
    {
        input: async (): Promise<Record<string, any>> => ({ a: true }),
        headers: [
            { a: true },
            { b: "custom:word" },
            { c: "faker:random:word" },
            { d: "resource:id" },
        ],
        output: true,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            b: await Helpers.word(),
        }),
        headers: [
            { a: true },
            { b: "custom:word" },
            { c: "faker:random:word" },
            { d: "resource:id" },
        ],
        output: true,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            c: faker.random.word(),
        }),
        headers: [
            { a: true },
            { b: "custom:word" },
            { c: "faker:random.word" },
            { d: "resource:id" },
        ],
        output: true,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            d: ValidResource.id,
        }),
        headers: [
            { a: true },
            { b: "custom:word" },
            { c: "faker:random:word" },
            { d: "resource:id" },
        ],
        output: true,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            d: ValidResource.id,
        }),
        headers: [
            { a: true, b: "custom:word" },
            { b: "custom:word", c: "faker:random:word" },
            { c: "faker:random:word", d: "resource:id" },
            { d: "resource:id", a: true },
        ],
        output: false,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            a: true,
            d: ValidResource.id,
            c: faker.random.word(),
            b: await Helpers.word(),
        }),
        headers: [
            { a: true, b: "custom:word" },
            { b: "custom:word", c: "faker:random:word" },
            { c: "faker:random:word", d: "resource:id" },
            { d: "resource:id", a: true },
        ],
        output: true,
    },
    {
        input: async (): Promise<Record<string, any>> => ({
            a: true,
            c: faker.random.word(),
        }),
        headers: [
            { a: true, b: "custom:word" },
            { b: "custom:word", c: "faker:random:word" },
            { c: "faker:random:word", d: "resource:id" },
            { d: "resource:id", a: true },
        ],
        output: false,
    },
];

const InvalidInputs: { input: any; headers: any[] }[] = [
    {
        input: {
            a: true,
        },
        headers: [{ a: "custom:invalidHelper" }],
    },
    {
        input: {
            a: true,
        },
        headers: [{ a: "custom:non.existent" }],
    },
    {
        input: {
            a: true,
        },
        headers: [{ a: "resource:addresses" }],
    },
    {
        input: {
            a: true,
        },
        headers: [{ a: "resource:lorem.ipsum" }],
    },
];
describe(`Header match`, () => {
    test.each(ValidInputs)(`Valid entry: %#`, async (entry) =>
        expect(
            headersValidation(
                await entry.input(),
                entry.headers,
                ValidResource,
                Helpers
            )
        ).resolves.toBe(entry.output)
    );
    test.each(InvalidInputs)(`Invalid entry: %#`, async (entry) =>
        expect(
            headersValidation(
                entry.input,
                entry.headers,
                ValidResource,
                Helpers
            )
        ).rejects.toMatchSnapshot()
    );
});
