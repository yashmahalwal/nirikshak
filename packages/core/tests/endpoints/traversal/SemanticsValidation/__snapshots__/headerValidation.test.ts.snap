// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header match Invalid entry: 0 1`] = `[Error: Non literal type value from custom:invalidHelper]`;

exports[`Header match Invalid entry: 1 1`] = `[Error: function: custom:non.existent, args: ]`;

exports[`Header match Invalid entry: 2 1`] = `[Error: Resource string resource:addresses resolved to non primitive value]`;

exports[`Header match Invalid entry: 3 1`] = `[Error: Invalid resource instance path: lorem.ipsum]`;
